src/theory/strings/output_channel.cpp

```diff
@@ -0,0 +1,448 @@
+    Trace("strings-lemma") << "Strings::Conflict : " << c << " : " << ant
+                           << std::endl;
+    Trace("strings-assert")
+        << "(assert (not " << ant << ")) ; conflict " << c << std::endl;
+    d_out.conflict(ant);
+    d_parent.d_conflict = true;
+    return;
+  }
+  Node lem;
+  if (ant == d_true)
+  {
+    lem = conc;
+  }
+  else
+  {
+    lem = NodeManager::currentNM()->mkNode(IMPLIES, ant, conc);
+  }
+  Trace("strings-lemma") << "Strings::Lemma " << c << " : " << lem << std::endl;
+  Trace("strings-assert") << "(assert " << lem << ") ; lemma " << c
+                          << std::endl;
+  d_pending_lem.push_back(lem);
+}
+
+void OutputChannelStrings::sendInfer(Node eq_exp, Node eq, const char* c)
+{
+  if (options::stringInferSym())
+  {
+    std::vector<Node> vars;
+    std::vector<Node> subs;
+    std::vector<Node> unproc;
+    inferSubstitutionProxyVars(eq_exp, vars, subs, unproc);
+    if (unproc.empty())
+    {
+      Node eqs =
+          eq.substitute(vars.begin(), vars.end(), subs.begin(), subs.end());
+      if (Trace.isOn("strings-lemma-debug"))
+      {
+        Trace("strings-lemma-debug") << "Strings::Infer " << eq << " from "
+                                     << eq_exp << " by " << c << std::endl;
+        Trace("strings-lemma-debug")
+            << "Strings::Infer Alternate : " << eqs << std::endl;
+        for (unsigned i = 0, nvars = vars.size(); i < nvars; i++)
+        {
+          Trace("strings-lemma-debug")
+              << "  " << vars[i] << " -> " << subs[i] << std::endl;
+        }
+      }
+      sendLemma(d_true, eqs, c);
+      return;
+    }
+    if (Trace.isOn("strings-lemma-debug"))
+    {
+      for (const Node& u : unproc)
+      {
+        Trace("strings-lemma-debug")
+            << "  non-trivial exp : " << u << std::endl;
+      }
+    }
+  }
+  Trace("strings-lemma") << "Strings::Infer " << eq << " from " << eq_exp
+                         << " by " << c << std::endl;
+  Trace("strings-assert") << "(assert (=> " << eq_exp << " " << eq
+                          << ")) ; infer " << c << std::endl;
+  d_pending.push_back(eq);
+  d_pending_exp[eq] = eq_exp;
+  d_keep.insert(eq);
+  d_keep.insert(eq_exp);
+}
+
+bool OutputChannelStrings::sendSplit(Node a, Node b, const char* c, bool preq)
+{
+  Node eq = a.eqNode(b);
+  eq = Rewriter::rewrite(eq);
+  if (eq.isConst())
+  {
+    return false;
+  }
+  NodeManager* nm = NodeManager::currentNM();
+  Node lemma_or = nm->mkNode(OR, eq, nm->mkNode(NOT, eq));
+  Trace("strings-lemma") << "Strings::Lemma " << c << " SPLIT : " << lemma_or
+                         << std::endl;
+  d_pending_lem.push_back(lemma_or);
+  sendPhaseRequirement(eq, preq);
+  return true;
+}
+
+void OutputChannelStrings::sendPhaseRequirement(Node lit, bool pol)
+{
+  d_pending_req_phase[lit] = pol;
+}
+
+Node OutputChannelStrings::getRepresentative(Node t)
+{
+  if (d_ee.hasTerm(t))
+  {
+    return d_ee.getRepresentative(t);
+  }
+  return t;
+}
```

Comment:
IMHO, we should only have an implementation here or in `TheoryStrings` and not in both.